### PR TITLE
Adds `__copy__` and `__deepcopy__` magic methods to `Immutable` class

### LINKED ISSUE #276: Our `Immutable` type should support `__copy__` and `__deepcopy__` methods
Currently our definition of `Immutable` misses these two methods:

```python
class Immutable(object):
    def __copy__(self):
        return self

    def __deepcopy__(self, memo):
        return self
```

### PR DESCRIPTION
**Pull request by phakiller**
The `__copy__` and `__deepcopy__` magic methods was implemented, both returning the `self` reference.

Closes #276

### CODE REVIEW

**Comment by phakiller** [comment]
@sobolevn I didn't write the tests yet, I don't know how strategy use to make them. We don't have any tests for "Immutable" class directly, so have I complement the tests of classes that are inherited from Immutable with `copy` and `deepcopy` functions?

**Comment by phakiller** [comment]
I've implemented the tests!

File: tests/test_context/test_requires_context/test_context.py [hunk 1]
```diff
@@ -21,6 +23,24 @@ def test_protocols(container, protocol):
...
 
 def test_context_immutable():
-    """Ensures that RequiresContext container supports ``.map()`` method."""
+    """Ensures that Context is immutable."""
     with pytest.raises(ImmutableStateError):
         Context().abc = 1
+
+
+def test_context_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_context/test_requires_context/test_context.py:32]: Can you please refactor this doctest to take a single line and remove `noqa`?
I think that there's no need for it here.

File: tests/test_context/test_requires_context/test_context.py [hunk 2]
```diff
@@ -21,6 +23,24 @@ def test_protocols(container, protocol):
...
+    Ensures that Context instance will returns it self
+    when passed to `copy` function.
+    """
+    context: Context = Context()
+    assert context is copy(context)
+
+
+def test_context_immutable_deepcopy():
+    """ # noqa
```
Comment (sobolevn) [inline tests/test_context/test_requires_context/test_context.py:41]: The same applies here.

File: tests/test_context/test_requires_context_io_result/test_context_io_result.py [hunk 3]
```diff
@@ -50,3 +52,23 @@ def test_requires_context_result_immutable():
     """Ensures that container is immutable."""
     with pytest.raises(ImmutableStateError):
         RequiresContextIOResult.from_success(1).abc = 1
+
+
+def test_requires_context_result_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_context/test_requires_context_io_result/test_context_io_result.py:58]: The same applies here.

File: tests/test_context/test_requires_context_io_result/test_context_io_result.py [hunk 4]
```diff
@@ -50,3 +52,23 @@ def test_requires_context_result_immutable():
...
+    Ensures that container instance will returns it self
+    when passed to `copy` function.
+    """
+    context_i_o_result: ContextIOResult = ContextIOResult()
+    assert context_i_o_result is copy(context_i_o_result)
+
+
+def test_requires_context_result_immutable_deepcopy():  # noqa: WPS118
+    """ # noqa
```
Comment (sobolevn) [inline tests/test_context/test_requires_context_io_result/test_context_io_result.py:67]: The same applies here.

File: tests/test_context/test_requires_context_result/test_context_result.py [hunk 5]
```diff
@@ -43,6 +45,23 @@ def test_context_result_immutable():
         ContextResult().abc = 1
 
 
+def test_context_result_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_context/test_requires_context_result/test_context_result.py:49]: The same applies here.

File: tests/test_context/test_requires_context_result/test_context_result.py [hunk 6]
```diff
@@ -43,6 +45,23 @@ def test_context_result_immutable():
...
+    Ensures that helper will returns it self when passed to `copy` function.
+    """
+    context_result: ContextResult = ContextResult()
+    assert context_result is copy(context_result)
+
+
+def test_context_result_immutable_deepcopy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_context/test_requires_context_result/test_context_result.py:57]: The same applies here.

File: tests/test_maybe/test_maybe_equality.py [hunk 7]
```diff
@@ -79,3 +81,39 @@ def test_immutability_success():
 
     with pytest.raises(AttributeError):
         Some(1).missing  # type: ignore # noqa: WPS428
+
+
+def test_success_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_maybe/test_maybe_equality.py:87]: The same applies here.

File: tests/test_maybe/test_maybe_equality.py [hunk 8]
```diff
@@ -79,3 +81,39 @@ def test_immutability_success():
...
+    Ensures that Success container instance will returns it self
+    when passed to `copy` function.
+    """
+    some: Maybe = Some(1)  # noqa: WPS110
+    assert some is copy(some)
+
+
+def test_success_immutable_deepcopy():
+    """ # noqa
```
Comment (sobolevn) [inline tests/test_maybe/test_maybe_equality.py:96]: The same applies here.

File: tests/test_maybe/test_maybe_equality.py [hunk 9]
```diff
@@ -79,3 +81,39 @@ def test_immutability_success():
...
     with pytest.raises(AttributeError):
         Some(1).missing  # type: ignore # noqa: WPS428
+
+
+def test_success_immutable_copy():
+    """  # noqa
+    Ensures that Success container instance will returns it self
+    when passed to `copy` function.
+    """
+    some: Maybe = Some(1)  # noqa: WPS110
```
Comment (sobolevn) [inline tests/test_maybe/test_maybe_equality.py:91]: Why do we need `noqa` here?

File: tests/test_maybe/test_maybe_equality.py [hunk 10]
```diff
@@ -79,3 +81,39 @@ def test_immutability_success():
...
+    """ # noqa
+    Ensures that Success container instance will returns it self
+    when passed to `deepcopy` function.
+    """
+    some: Maybe = Some(1)  # noqa: WPS110
+    assert some is deepcopy(some)
+
+
+def test_failure_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_maybe/test_maybe_equality.py:105]: The same applies here.

File: tests/test_maybe/test_maybe_equality.py [hunk 11]
```diff
@@ -79,3 +81,39 @@ def test_immutability_success():
...
+    Ensures that Failure container instance will returns it self
+    when passed to `copy` function.
+    """
+    nothing: Maybe = _Nothing()
+    assert nothing is copy(nothing)
+
+
+def test_failure_immutable_deepcopy():
+    """ # noqa
```
Comment (sobolevn) [inline tests/test_maybe/test_maybe_equality.py:114]: The same applies here.

File: tests/test_result/test_result_equality.py [hunk 12]
```diff
@@ -82,3 +84,39 @@ def test_immutability_success():
 
     with pytest.raises(AttributeError):
         Success(1).missing  # type: ignore # noqa: WPS428
+
+
+def test_success_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_result/test_result_equality.py:90]: The same applies here.

File: tests/test_result/test_result_equality.py [hunk 13]
```diff
@@ -82,3 +84,39 @@ def test_immutability_success():
...
+    Ensures that Success container instance will returns it self
+    when passed to `copy` function.
+    """
+    success: Result = Success(1)
+    assert success is copy(success)
+
+
+def test_success_immutable_deepcopy():
+    """ # noqa
```
Comment (sobolevn) [inline tests/test_result/test_result_equality.py:99]: The same applies here.

File: tests/test_result/test_result_equality.py [hunk 14]
```diff
@@ -82,3 +84,39 @@ def test_immutability_success():
...
+    Ensures that Success container instance will returns it self
+    when passed to `deepcopy` function.
+    """
+    success: Result = Success(1)
+    assert success is deepcopy(success)
+
+
+def test_failure_immutable_copy():
+    """  # noqa
```
Comment (sobolevn) [inline tests/test_result/test_result_equality.py:108]: The same applies here.

File: tests/test_result/test_result_equality.py [hunk 15]
```diff
@@ -82,3 +84,39 @@ def test_immutability_success():
...
+    Ensures that Failure container instance will returns it self
+    when passed to `copy` function.
+    """
+    failure: Result = Failure(0)
+    assert failure is copy(failure)
+
+
+def test_failure_immutable_deepcopy():
+    """ # noqa
```
Comment (sobolevn) [inline tests/test_result/test_result_equality.py:117]: The same applies here.

**Comment by phakiller** [comment]
> Thanks a lot for your work!
> 
> Can you please fix my minor comments before I am merging this?

Done

File: tests/test_result/test_result_equality.py [hunk 16]
```diff
@@ -82,3 +84,27 @@ def test_immutability_success():
 
     with pytest.raises(AttributeError):
         Success(1).missing  # type: ignore # noqa: WPS428
+
+
+def test_success_immutable_copy():
+    """Ensures that Success returns it self when passed to copy function."""
+    success: Result = Success(1)
```
Comment (sobolevn) [inline tests/test_result/test_result_equality.py:91]: Do we need `: Result` part here?

Comment (phakiller) [reply tests/test_result/test_result_equality.py:91]: Maybe don't, I'll try without this part!

**Comment by sobolevn** [comment]
Thanks a lot!

**Pull request merged**